BPF MASQ for veth, ip-masq-agent and multi-dev NodePort

brb opened the pull request:
**DON'T MERGE BEFORE https://github.com/cilium/cilium/pull/11148 GETS MERGED**

This PR introduces the following features:

- BPF masquerading for veth mode (ipv4-only).
- BPF ip-masq-agent.
- Multi-dev support for NodePort and BPF masq.
- Removes BPF masq from ipvlan.

Currently, the BPF masq depends on NodePort. I'm planning to decouple both, but this requires some big refactoring which I was hesitant to do considering the size of this PR.

Updates to documentation and a couple of items mentioned in the PR comments by me will be added in a follow-up.

Reviewable per commit.

Fix #10229 
Fix #9922 
Fix #9620

```release-note
Add BPF masquerading for veth mode
Allow attaching BPF NodePort to multiple devices
Add BPF-based ip-masq-agent
```

brb committed:
datapath: Enable BPF MASQ for veth mode in IPv4

This commit enables MASQUERADE'ing packets sent from local endpoints to
outside (WORLD_ID).

The MASQ is enabled only when --enable-bpf-masquerade is set.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Remove ipvlan MASQ and dead code

Currently, it is broken, and it requires a re-work. The removal will
help us to refactor the BPF SNAT engine.

Also, this allows us to remove the SNAT dead code, and to inline the
helper macros.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Add --enable-bpf-masquerade flag

Currently, the flag is noop except that it disables provisioning of
iptables MASQUERADE rules when set.

The flag is going to be used to control BPF MASQUERADE'ing.

By default, it is set to false, so that users who will upgrade to
Cilium v1.8 will keep the existing behaviour (= MASQ is done by
iptables).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Fix BPF MASQ for veth when DSR is enabled

This commit fixes DSR and BPF MASQ when both are enabled:

- For egress path (local_endpoint -> outside) do not skip (S)NAT if L4
  protocol uses DSR.
- For ingress do not avoid trying to do the reverse SNAT xlation if L4
  protocol uses DSR. Currently, the rev-xlaton happens inside the
  CILIUM_CALL_IPV4_NODEPORT_NAT tailcall.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm: Add bpfMasquerade param

The param controls whether BPF MASQ should be enabled. Currently, it
defaults to "false", and cilium-agent defaults to "false" as well, so
that existing users who upgrade to v1.8 will preserve the existing
behavior (iptables).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Add BPF and ip-masq-agent integration tests

Also, move echoserver-hostnetns into DaemonSet (previously, in k8s v1.11
the echoserver pod could not be scheduled on the node which didn't run cilium,
regardless that the pod was tolerating all taints).

Moving pod into the DaemonSet fixes the problem.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Add ip-masq-agent flags

This commit introduces the following ip-masq-agent related flags:

* --enable-ip-masq-agent
* --ip-masq-agent-config-path
* --ip-masq-agent-sync-period

Currently, the flags are noop.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
maps/ipmasq: Add LPM CIDR map for ip-masq-agent

The LPM map is going to be used by Cilium's ip-masq-agent to
store CIDRs which should bypass masquerading.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
ipmasq: Add ip-masq-agent package

This commit introduces the "ip-masq-agent" controller which is going to
be used to sync the ipmasq BPF maps with a config files updated by
users.

Also, enable ip-masq-agent when --enable-ip-masq-agent is set to true.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
ipmasq: Configure via YAML instead of JSON

It makes an integration via ConfigMap much easier, as a user provided
ConfigMap will be mounted as a file to cilium-agent pod, and its content
is encoded as YAML.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
helm: Add ipMasqAgent option

The BPF ip-masq-agent can be enabled with global.ipMasqAgent.enabled.

When enabled, the ip-masq-agent configmap (created by a user) is mounted
into /etc/config/ip-masq-agent of the cilium-agent pod.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Replace ct_entry.backend_id with ifindex

The field is no longer used, and we can use it to store ifindex of a
NodePort device. This will be needed when we will run bpf_netdev.o on
multiple devices.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
common: Move GoArray2C to common

It's going to be used by multiple packages.

Also, add GoArray2CNoSpaces.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Add secondary private network to CI

The network is reachable via enp0s9 in VMs, and it's going to be used
to test NodePort when bpf_netdev.o is attached to multiple devices.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Add NodePort tests for multiple devices

The test cases are skipped when running in other CNI integrations
(EKS,GKE) as the secondary device might not be available.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
cli: Cleanup multiple bpf_netdev.o programs upon cleanup

Instead of relying on NATIVE_DEV_IFINDEX from node_config.h which no
longer exists, filter all devices and check whether they have TC
ingress/egress BPF filter which name is bpf_netdev.o.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Check NodePort --device names and ifindices

Check that ifindex of a given device does not overflow uin16, and that
the device name does not contain special symbols used to parse in
bpf/init.sh.

The first restriction will go away once we introduce the surrogate id =>
ifindex BPF map. While the second - when we migrate bpf/init.sh to Go.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
cli: Add NodePort devices to cilium status

The example of output:

```
$ cilium status | grep KubeProxyReplacement

KubeProxyReplacement:   Strict   (eth0, eth1)   [NodePort (HYBRID, 30000-32767, XDP: NONE), ExternalIPs, HostReachableServices (TCP, UDP)]
```

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
probe: Set RLIMIT_MEMLOCK=RLIM_INFINITY for HaveFullLPM

Otherwise, the map creation in the probe (sometimes) fails with EPERM.
See [1] for details re rlimit.

[1]: https://docs.cilium.io/en/v1.7/bpf/#miscellaneous

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
test: Extend testNodePort

Extend the test function to accept a param which indicates whether
tests from a third (aka node w/o cilium) node should be performed.

As a reminder, the third node is used for NodePort testing to avoid
bpf_sock.c optimizations which happen when a request is sent to
a NodePort svc from a node which runs cilium.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
cli: Add cilium bpf ipmasq list

The cmd lists all CIDRs in the ip-masq-agent BPF LPM map.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Warn about rp_filter=1 in DSR mode

In the direct routing mode, in order for DSR to work via non-direct
routing iface, a user should set rp_filter either to 0 or to 2 for the
direct routing iface.

Otherwise, the kernel might drop a forwarded packet due to the strict
source IP addr check.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb commented on pkg/maps/ipmasq/ipmasq.go:
```diff
@@ -0,0 +1,100 @@
+
+const (
+	MapName    = "cilium_ipmasq_v4"
+	maxEntries = 16384
```
Note to myself: make it configurable.

joestringer replied:
And maybe add to the docs page that lists all of the map limits?

brb commented on daemon/cmd/daemon_main.go:
```diff
@@ -1521,7 +1554,46 @@ func initKubeProxyReplacementOptions() {
+				log.Fatalf("%s link name contains ',', '=' or ';' character which is not allowed")
+			}
+			if idx := link.Attrs().Index; idx > math.MaxUint16 {
+				// TODO(brb) after adding surrogate id => ifindex BPF map,
```
This will be added as a follow-up.

brb commented on bpf/lib/nat.h:
```diff
@@ -126,6 +126,17 @@ struct bpf_elf_map __section_maps SNAT_MAPPING_IPV4 = {
+	.size_key	= sizeof(struct lpm_v4_key),
+	.size_value	= sizeof(struct lpm_val),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 16384,
```
Note to myself: make it configurable in a follow-up.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -510,6 +508,9 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
 	case CT_NEW:
 		ct_state_new.src_sec_id = SECLABEL;
 		ct_state_new.node_port = 1;
+#ifdef NATIVE_DEV_IFINDEX
```
All these #ifdef's are required to compile bpf_lxc.c, as it depends on nodeport.h. Alternatively, I could define a dummy value in lxc_config.h.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1251,64 +1294,36 @@ static __always_inline int nodeport_nat_fwd(struct __ctx_buff *ctx,
 			addr = IPV4_GATEWAY;
 		else
 #endif
+#ifdef IPV4_NODEPORT
```
Again, this #ifdef is needed to compile bpf_lxc.c.

borkmann commented on bpf/init.sh:
```diff
@@ -456,11 +456,17 @@ case "${MODE}" in
+			MAC=$(mac2array $MAC)
+			MAC_BY_IFINDEX_MACRO="${MAC_BY_IFINDEX_MACRO}	case ${IDX}: {union macaddr tmp = {.addr = ${MAC}}; mac=tmp;} break; \\\\\n"
+			MAC_BY_IFINDEX_MACRO="${MAC_BY_IFINDEX_MACRO}${MAC_BY_IFINDEX_MACRO_END}"
+			echo -e "${MAC_BY_IFINDEX_MACRO}" >> $RUNDIR/globals/node_config.h
```
Nice idea, Did you check whether LLVM generates a search-tree on this construct if there are say >=8 entries?

Given the number of such devices is usually not too high that would have outside world connectivity, I wonder whether we could do the same trick to overcome the u16 ifindex value problem?

We could have code-generation for two macros (or __always_inline functions), i) that is generating the following. Pseudo code for 2 devices in hostns (one with ifindex 42, another one with 147483647):

```
static __always_inline __u8 ifindex_to_dev_id(const int ifindex)
{
    switch (ifindex) {
    case 42: return 1;
    case 147483647: return 2;
    default: return 0; // invalid one
    }
}

static __always_inline int dev_id_to_ifindex(const __u8 dev_id)
{
    switch (dev_id) {
    case 1: return 42;
    case 2: return 147483647;
    default: return 0; // invalid one
    }
}
```

If we set a limit of max 255 such devices that we track at the same time, they can have whatever ifindex they'd want, but we can store the tracking more efficiently in the CT with just 8 bits.

It would work since we know the full range of devices we need to attach to a prior anyway since it's done in init.sh.

Wdyt?

brb replied:
> Did you check whether LLVM generates a search-tree on this construct if there are say >=8 entries?

Noup, I haven't, but I'll do it.

> Wdyt?

Yeah, sounds like a solution. Just we need to make sure that after cilium-agent restarts with a new configuration previously used devices get the same surrogate ids.

Also, I'd like not to complicate this PR, and to do it in a follow-up PR.

borkmann replied:
Yep, sounds good, lets keep tracking these items somewhere then for follow-up.

borkmann commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -45,7 +45,7 @@ cilium-agent [flags]
   -D, --debug                                         Enable debugging mode
       --debug-verbose strings                         List of enabled verbose debug groups
-  -d, --device string                                 Device facing cluster/external network for direct L3 (non-overlay mode) (default "undefined")
+  -d, --device strings                                List of devices facing cluster/external network for attaching bpf_netdev
```
Could we let users also specify a device name regex? E.g. some nodes may have different device names than others when provisioning, but they might all share a same common prefix for public interface e.g. `em*`, `eth*` or `pub*` (if configured as such to contrast the `lxc*` prefix) where the daemon would then internally gather the concrete list we need to attach to?

brb replied:
Yeah, that's a good idea. I will implement it in a follow-up.

brb committed:
test: Allow to specify sport in testDSR()

Re-using the same port might lead to curl failing to bind, as the port
might be still not released (exit code 45).

Therefore, allow to specify a custom source port.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann commented on Documentation/gettingstarted/ipvlan.rst:
```diff
@@ -107,20 +101,6 @@ masquerading all traffic leaving the node:
-     --set global.masquerade=true \\
-     --set global.installIptablesRules=false \\
-     --set global.autoDirectNodeRoutes=true
-
```
Do you plan to re-add the documentation here on how to use ipvlan in L3-only mode with the ip masq agent?

brb replied:
Yep, but only after we have tested that the ipvlan masq actually works (in a follow up).

borkmann commented on bpf/lib/common.h:
```diff
@@ -665,6 +667,21 @@ static __always_inline int redirect_peer(int ifindex __maybe_unused,
 
+struct lpm_v4_key {
+		struct bpf_lpm_trie_key lpm;
+			__u8 addr[4];
```
nit: indent

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -325,7 +313,9 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
-	fib_params.ifindex = NATIVE_DEV_IFINDEX;
+#ifdef DIRECT_ROUTING_DEV_IFINDEX
+	fib_params.ifindex = DIRECT_ROUTING_DEV_IFINDEX;
+#endif
```
Is this "first device in the --device list being the direct routing one" due to the above? Is the helper required to haver some ifindex as input? Could we make this more generic in that the user doesn't need to keep this in mind?

brb replied:
> Is this "first device in the --device list being the direct routing one" due to the above?

Yes, currently we don't determine which iface from the --device list is used for the direct routing. In a follow-up, I will explore whether k8s Node IP addr is available before exec'ing bpf/init.sh. If yes, then we can dynamically determine such an iface, and thus the requirement for the first device in the list could be lifted.

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -704,51 +707,78 @@ static __always_inline bool nodeport_uses_dsr4(const struct ipv4_ct_tuple *tuple
+#ifdef ENABLE_IP_MASQ_AGENT
+				struct lpm_v4_key pfx;
+				pfx.lpm.prefixlen = 32;
+				__builtin_memcpy(pfx.lpm.data, &ip4->daddr, sizeof(pfx.addr));
```
nit: just memcpy()

borkmann commented on bpf/lib/nodeport.h:
```diff
@@ -1057,14 +1096,10 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		ctx_store_meta(ctx, CB_NAT, NAT_DIR_INGRESS);
+		ctx_store_meta(ctx, CB_SRC_IDENTITY, src_identity);
+		ep_tail_call(ctx, CILIUM_CALL_IPV4_NODEPORT_NAT);
+		return DROP_MISSED_TAIL_CALL;
```
Do we also need this if ip masq is disabled and we run with full DSR?

brb replied:
Good spot, will add #ifdef.

borkmann commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -295,6 +319,21 @@ var _ = Describe("K8sDatapathConfig", func() {
+				"global.bpfMasquerade":        "true",
+				"global.nodePort.device":      fmt.Sprintf(`'{%s,%s}'`, privateIface, defaultIface),
+				"global.tunnel":               "disabled",
+				"global.autoDirectNodeRoutes": "true",
```
Could we also add a test when XDP acceleration is enabled as well to make sure both don't break when enabled at the same time?

borkmann replied:
Btw, should we get the small fix in #10874 resolved first to make sure we have proper CI coverage for the new tests on 4.19?

brb replied:
Re testing with XDP: how flaky is the XDP test? Maybe worth waiting until #10942 is resolved to avoid adding new flakes?

Re 4.19: if #10874 gets merged soon (hopefully the last commit has fixed the issue), then yeah, we can wait. Anyway, I'd prefer to get this PR (#10878) merged sooner than later, as rebasing && making sure that CI is green is taking too much time and effort.

aanm commented on bpf/init.sh:
```diff
@@ -465,14 +460,19 @@ case "${MODE}" in
+				MAC_BY_IFINDEX_MACRO="${MAC_BY_IFINDEX_MACRO}	case ${IDX}: {union macaddr tmp = {.addr = ${MAC}}; mac=tmp;} break; \\\\\n"
+			done
+			MAC_BY_IFINDEX_MACRO="${MAC_BY_IFINDEX_MACRO}${MAC_BY_IFINDEX_MACRO_END}"
+			echo -e "${MAC_BY_IFINDEX_MACRO}" >> $RUNDIR/globals/node_config.h
```
I wonder how much of this logic could exist in go. (Obviously could be addressed in a follow up)

brb replied:
Perhaps the whole bpf/init.sh could be replaced in Go (AFAIK there was an issue for this, can't find it though).

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -251,7 +253,7 @@ func init() {
 	option.BindEnv(option.DebugVerbose)
 
-	flags.StringP(option.Device, "d", "undefined", "Device facing cluster/external network for direct L3 (non-overlay mode)")
+	flags.StringSliceP(option.Device, "d", []string{}, "List of devices facing cluster/external network for attaching bpf_netdev (first device should be one used for direct routing if tunneling is disabled)")
```
`first device should be one used for direct routing if tunneling is disabled` what if the packets are supposed to be routed to the 2nd device's network?

brb replied:
Then that device needs to be first in the list. See https://github.com/cilium/cilium/pull/10878#discussion_r406926434 for more details.

aanm replied:
I don't think you got my point. Sure I add the 2nd device-B as first on the list, now what happens to packets that are supposed to be routed to the device-A that is 2nd on the list?

brb replied:
This PR doesn't add a support for direct routing via multiple devices.

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1553,22 +1586,63 @@ func initKubeProxyReplacementOptions() {
 		}
 	}
+	if option.Config.EnableNodePort {
+		for _, iface := range option.Config.Devices {
+			link, err := netlink.LinkByName(iface)
+			if err != nil {
```
This could live in a function

brb replied:
I'd prefer to have it as it is - moving to a function might hinder readability of `initKubeProxyReplacementOptions()`.

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -1553,22 +1586,63 @@ func initKubeProxyReplacementOptions() {
+			}
+		}
+
+		if option.Config.EnableIPv4 &&
+			option.Config.Tunnel == option.TunnelDisabled &&
+			option.Config.NodePortMode != option.NodePortModeSNAT &&
+			len(option.Config.Devices) > 1 {
```
Add a comment stating what is this logic doing.

brb replied:
Isn't the warn statement in the if-body informative enough?

aanm replied:
It's not clear to me what this conditions have to do with the warning bellow.

brb replied:
Added the comment. I hope that now it's more clear.

aanm commented on install/kubernetes/cilium/charts/config/templates/configmap.yaml:
```diff
@@ -309,7 +314,7 @@ data:
 {{- end }}
 {{- if .Values.global.nodePort.device }}
-  device: {{ .Values.global.nodePort.device | quote }}
+  device: {{ join " " .Values.global.nodePort.device | quote }}
```
Why the `join " "` here if `.Values.global.nodePort.device` is suppose to be a `single` device?

brb replied:
The PR adds a possibility to specify multiple devices. For now, I didn't want to introduce a new helm option for that, and re-used the existing one. After this PR gets merged, we can discuss during the weekly meeting whether a new option should be added.

aanm replied:
It's cleaner we if we add a new helm variable and deprecate the old one.

brb replied:
Yes, but let's do it in a follow-up.

aanm commented on install/kubernetes/quick-install.yaml:
```diff
@@ -161,6 +148,24 @@ data:
+---
+# Source: cilium/charts/agent/templates/serviceaccount.yaml
+apiVersion: v1
+kind: ServiceAccount
```
Please update your local helm and regenerate these files.

aanm commented on pkg/datapath/loader/base.go:
```diff
@@ -251,8 +250,32 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
+					fmt.Sprintf("%s=%#x", iface,
+						byteorder.HostSliceToNetwork(ipv4, reflect.Uint32).(uint32)))
+			}
+			args[initArgNodePortIPv4Addrs] = strings.Join(tmp, ";")
```
It seems inconsistent that in here we are joining iface names with `;` but on line 236 we are doing with `,`

brb replied:
(We cannot join the ipv6 addr with `,`, as each addr contains `,`). Replaced 236 with `;`.

aanm commented on pkg/maps/ctmap/types.go:
```diff
@@ -475,7 +475,7 @@ type CtEntry struct {
 	// RevNAT is in network byte order
 	RevNAT           uint16 `align:"rev_nat_index"`
-	_                uint16 `align:"backend_id"`
+	IfIndex          uint16 `align:"ifindex"`
```
`ifindex` -> `if_index`

brb replied:
It's `ifindex`.

aanm replied:
It seems inconsistent given that we have `_` between words but not fox this variable.

brb replied:
The datapath code is using `ifindex` everywhere.

aanm commented on pkg/node/node_address.go:
```diff
@@ -37,10 +37,10 @@ var (
 	ipv4ExternalAddress net.IP
 	ipv4InternalAddress net.IP
-	ipv4NodePortAddress net.IP
+	ipv4NodePortAddrs   map[string]net.IP // iface => ip addr
```
nit `// iface => ip addr` -> `// iface name => ip addr`

aanm commented on pkg/node/node_address.go:
```diff
@@ -37,10 +37,10 @@ var (
 	ipv6Address         net.IP
 	ipv6RouterAddress   net.IP
-	ipv6NodePortAddress net.IP
+	ipv6NodePortAddrs   map[string]net.IP // iface => ip addr
```
nit `// iface => ip addr` -> `// iface name => ip addr`

aanm commented on pkg/node/node_address.go:
```diff
@@ -203,24 +214,36 @@ func Uninitialize() {
+//
+// NOTE: Do not modify the returned map.
+func GetNodePortIPv4AddrsWithDevices() map[string]net.IP {
+	return ipv4NodePortAddrs
```
why does this function have this comment but `GetNodePortIPv4Addrs()` was copied before returning?a

aanm commented on pkg/node/node_address.go:
```diff
@@ -203,24 +214,36 @@ func Uninitialize() {
+// GetNodePortIPv4AddrsWithDevices returns the map iface => NodePort IPv6.
+//
+// NOTE: Do not modify the returned map.
+func GetNodePortIPv6AddrsWithDevices() map[string]net.IP {
```
why does this function have this comment but `GetNodePortIPv6Addrs()` was copied before returning?a

aanm commented on test/k8sT/manifests/externalIPs/README.md:
````diff
@@ -62,4 +62,4 @@ that run on `k8s1` and `k8s2`:
 ```
 
-and start cilium with `--enable-node-port --device=enp0s9`
+and start cilium with `--enable-node-port --device=enp0s10`
````
nit: what happened to `enp0s9`? Can't `enp0s10` be used for `ip-masq` tests instead?

brb replied:
For the multi-dev NodePort tests, I've introduced an additional private network (reachable via `enp0s9`), so the previous `enp0s9` was renamed to `enp0s10`.

joestringer commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -46,20 +46,22 @@ cilium-agent [flags]
   -D, --debug                                         Enable debugging mode
       --debug-verbose strings                         List of enabled verbose debug groups
-  -d, --device string                                 Device facing cluster/external network for direct L3 (non-overlay mode) (default "undefined")
+  -d, --device strings                                List of devices facing cluster/external network for attaching bpf_netdev (first device should be one used for direct routing if tunneling is disabled)
```
Maybe an aside from this PR as a whole, but where does this `(first device should be one used for direct routing if tunneling is disabled)` restriction come from? Why can't we just detect the default route to determine which device that is?

brb replied:
Not a default route, but a device with NodeIP. Please see https://github.com/cilium/cilium/pull/10878#discussion_r406926434.

The restriction comes from the fact that we cannot forward a packet to a remote node via a device which it came to if it doesn't have a podCIDR route.

joestringer commented on common/utils.go:
```diff
@@ -50,6 +50,40 @@ func C2GoArray(str string) []byte {
 	return ret
 }
 
+// GoArray2C transforms a byte slice into its hexadecimal string representation.
```
Why are we moving more code into `common/`? AFAIK we're trying to get rid of this directory.

brb replied:
Oops, didn't know that. In a follow-up I will move it to `pkg/utils` or similar.

joestringer commented on pkg/probe/probe.go:
```diff
@@ -44,6 +44,21 @@ func (p *probeValue) DeepCopyMapValue() bpf.MapValue { return &probeValue{p.Valu
+	if err := unix.Setrlimit(unix.RLIMIT_MEMLOCK, &tmpLim); err != nil {
+		return false
+	}
+	defer unix.Setrlimit(unix.RLIMIT_MEMLOCK, &oldLim)
```
We already `bpf.ConfigureResourceLimits()` in `daemon/cmd/daemon.go`. Is that not early enough?

brb replied:
Yep, it's not early enough - I call the probe in `initEnv()`, while `bpf.ConfigureResourceLimits()` is called in `runDaemon()` which happens after `initEnv()`.

Anyway, in a follow-up I will check whether it's possible to call `bpf.ConfigureResourceLimits()`  a bit earlier.

joestringer commented on pkg/ipmasq/ipmasq_test.go:
```diff
@@ -0,0 +1,143 @@
+func (i *IPMasqTestSuite) TestUpdate(c *check.C) {
+	_, err := i.configFile.WriteString("nonMasqueradeCIDRs:\n- 1.1.1.1/32\n- 2.2.2.2/16")
+	c.Assert(err, check.IsNil)
+	time.Sleep(300 * time.Millisecond)
```
What are we waiting for here? There's several sleeps in these unit tests which we should avoid if at all possible.

brb replied:
These are needed to wait until ipmasq goroutine has performed an update.

joestringer replied:
Could we check for the update using a channel instead?

brb replied:
Yeah, we could, but that would require changing the ipmasq implementation (allowing to pass a callback or so for testing only). What is your main worry re the sleeps? Is it that the test might become flaky or the test duration?

joestringer replied:
yes, both of these.

This is one of those 'death by a thousand cuts' things where we sprinkle a few seconds of sleeps here, a few seconds of sleeps there, and before long you spend a bunch of the time of your unit tests on waiting for the clock to tick.

I'm not *so* worried about injecting callbacks/interfaces in the real implementation if the operations aren't overly frequent.

brb replied:
Yeah, makes sense. Added to the follow-up list.

joestringer commented on pkg/ipmasq/ipmasq.go:
```diff
@@ -0,0 +1,176 @@
+	manager.UpdateController("ip-masq-agent",
+		controller.ControllerParams{
+			DoFunc: func(ctx context.Context) error {
+				return a.Update()
```
Did you consider signing up for filesystem events for update of the file rather than polling on a regular basis?

I guess it shouldn't matter that much though, I presume we expect these to be relatively small and not take a lot of time to process.

brb replied:
Yep, an ip-masq-agent config file is usually very simple and small, so therefore I was reluctant to use fsnotify.

joestringer reviewed (commented):
I did a first pass glance through a random scattering of the changes. I didn't look at all at the core implementation yet but I feel like I have some idea what it's trying to achieve at least which should help for next time. A few minor general maintenance nits below that you can consider adding to your list of followup items.

joestringer commented on daemon/cmd/daemon_main.go:
```diff
@@ -1056,34 +1071,45 @@ func initEnv(cmd *cobra.Command) {
-			log.WithField(logfields.Device, option.Config.Device).
-				Fatal("BPF masquerade needs external facing device specified")
+		if option.Config.EgressMasqueradeInterfaces != "" {
+			log.Fatalf("BPF masquerade does not allow to specify devices via --%s", option.EgressMasqueradeInterfaces)
```
Is there an alternative we can suggest, such as "try `--device` instead"?

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -708,51 +711,78 @@ static __always_inline bool nodeport_uses_dsr4(const struct ipv4_ct_tuple *tuple
+			return false;
+
+		// TODO(brb) use skb mark to determine if src is a local endpoint
+		if (__lookup_ip4_endpoint(ip4->saddr) != NULL) {
```
I believe this will return entries for host IPs too (`ep->flags & ENDPOINT_F_HOST`) though I guess if this is the `cilium_host` IP then it may still needs SNAT anyway. For native device IPs not though?

Here's my entries from dev VM for reference:

```
$ sudo cilium bpf endpoint list
IP ADDRESS                    LOCAL ENDPOINT INFO
fc00::10ca:1:0                (localhost)
10.0.2.15:0                   (localhost)
10.11.6.24:0                  (localhost)
10.11.24.144:0                id=2046  flags=0x0000 ifindex=14  mac=C2:FE:85:C5:E9:1F nodemac=C2:CD:F5:08:ED:5B
fe80::d0c1:f1ff:fea3:fdfc:0   (localhost)
f00d::a0b:0:0:231c:0          (localhost)
f00d::a0b:0:0:9f6d:0          id=2046  flags=0x0000 ifindex=14  mac=C2:FE:85:C5:E9:1F nodemac=C2:CD:F5:08:ED:5B
192.168.34.11:0               (localhost)
192.168.33.11:0               (localhost)
```

brb replied:
Hmm, good point. I think we should only MASQ outgoing connections from endpoints (changed). Do you know a case where a packet egressing to the world would have `cilium_host` IP as src?

joestringer replied:
I don't know any such case. `cilium_host` IP should only be chosen for source when traffic from the host is routed towards the `cilium_host` device, which is typically only the CIDR prefixes that Cilium manages.

The kind of case I'd be concerned about is something like if we have a virtual IP / DNAT type feature and set up routes for those destinations. However we don't route services to `cilium_host` anyway AFAIK, we will only do such service translation from the host-reachable services socket layer hooks. So that should not occur.

joestringer commented on bpf/lib/nodeport.h:
```diff
@@ -708,51 +711,78 @@ static __always_inline bool nodeport_uses_dsr4(const struct ipv4_ct_tuple *tuple
+					return true;
+#ifdef ENCAP_IFINDEX
+				if (info->sec_label == REMOTE_NODE_ID)
+					return true;
```
From the commit message:

> In the tunnel mode, a packet from local endpoint to remote node is not encapsulated and is sent via a native device.

My understanding was that the IP associated with `cilium_host` on the remote node in tunnel mode may be allocated by Cilium, and traffic for that IP routed via the tunnel, and Cilium would happily deliver it. I don't have a readily available multi-node cluster on hand during review right now to validate this though.

I'm guessing it's probably safe-ish to over-enable SNAT here, we'll just do more masquerading than necessary but it'd be interesting to follow up on whether this is always required for IPs associated with a remote-node and whether we may need additional info to determine how to skip this masquerade step for those IPs.

joestringer replied:
Oh, I see, the immediately following commit handles this case by passing down an `encap` bool from the callchain leading to this function (from `to_netdev` / `to_overlay`) which would break out before we hit this logic. LGTM, though might not be all that obvious how this case is handled by just reading this part of the code since it would imply that SNAT occurs. but *shrug*.

brb replied:
I'm adding a comment to make it more clear.

joestringer commented on pkg/maps/ipmasq/ipmasq.go:
```diff
@@ -0,0 +1,100 @@
+func key(cidr net.IPNet) *Key4 {
+	ones, _ := cidr.Mask.Size()
+	key := &Key4{PrefixLen: uint32(ones)}
+	copy(key.Address[:], cidr.IP.To4())
```
Is anything validating, for instance from the config file parsing -> `Update()` -> `key()`, that `cidr` here is IPv4 and that the `To4()` call here won't return `nil`?

brb replied:
Good point, I've added the validation to the caller (`pkg/ipmasq/ipmasq.go:(c *Ipnet) UnmarshalYAML()`), so that errors could be reported earlier.

joestringer commented on pkg/ipmasq/ipmasq.go:
```diff
@@ -0,0 +1,176 @@
+// IPMasqAgent represents a state of the ip-masq-agent
+type IPMasqAgent struct {
+	configPath             string
+	nonMasqCIDRsFromConfig map[string]net.IPNet
+	nonMasqCIDRsInMap      map[string]net.IPNet
```
nit, in other parts of the code we denote the difference here with:

```suggestion
	desiredNonMasqCIDRs map[string]net.IPNet
	realizedNonMasqCIDRs map[string]net.IPNet
```

brb replied:
Added to the follow-up list.

joestringer commented on pkg/ipmasq/ipmasq_test.go:
```diff
@@ -0,0 +1,143 @@
+	i.ipMasqMap = &ipMasqMapMock{cidrs: map[string]net.IPNet{}}
+	i.manager = controller.NewManager()
+
+	configFile, err := ioutil.TempFile("/tmp", "ipmasq-test")
```
There's no need to specify the tmp directory according to the ioutil API documentation:

```suggestion
	configFile, err := ioutil.TempFile("", "ipmasq-test")
```

joestringer commented on pkg/ipmasq/ipmasq.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+// IPMasqMap is an interface describing methods for manipulating an ipmasq map
+type IPMasqMap interface {
```
btw, awesome to see these abstractions and injection of the implementation + unit tests. :party-parrot:

joestringer commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -337,7 +375,113 @@ var _ = Describe("K8sDatapathConfig", func() {
+
+			// Wait until the ip-masq-agent configmap is mounted into cilium-agent pods,
+			// and the pods have read the new configuration
+			time.Sleep(90 * time.Second)
```
:/

I take it that this is done not because of the 1s reaction time we've configured in cilium-agent for changes to ip-masq, but rather because we're relying on the periodic sync of the configmap into the container which may take quite some time.

I would love to see a better approach for the validation, there's three minutes here alone.

brb replied:
I think I can poll the ip-masq-agent BPF map entries to detect the update. Added to the follow-up.

joestringer replied:
Even if we poll rather than sleep, I'm guessing that it still may take around this amount of time because we're waiting on the configmap to be updated into the container, no? @aanm could probably tell you how long it takes for a configmap change to become effective inside a container's filesystem.

aanm replied:
30 - 60 seconds I would say. 0 seconds if the pod is restarted

joestringer commented on bpf/lib/conntrack.h:
```diff
@@ -766,19 +746,13 @@ static __always_inline int ct_create4(const void *map_main,
 	 * the dir=CT_SERVICE (see GH#7060). Therefore, we can safely abuse
-	 * this field to save the backend_id. The hack will go away once we stop
-	 * supporting the legacy svc (in v1.6 we will zero the backend_id
-	 * field, in v1.7 we can remove the rx_bytes hack).
```
I guess we ended up sticking with the rx_bytes hack to avoid resizing the conntrack entries...?

brb replied:
Unfortunately, yes, as resizing the conntrack entries is not straightforward at all.

joestringer commented on bpf/init.sh:
```diff
@@ -471,14 +466,19 @@ case "${MODE}" in
-			NATIVE_DEV_MAC=$(mac2array $NATIVE_DEV_MAC)
-			echo "#define NATIVE_DEV_MAC { .addr = ${NATIVE_DEV_MAC}}" >> $RUNDIR/globals/node_config.h
-
+			MAC_BY_IFINDEX_MACRO="#define NATIVE_DEV_MAC_BY_IFINDEX(IFINDEX) ({ \\
```
I guess we won't support hot-plugging these.

(It's in theory supportable, someone would just need to convince `cilium-agent` to fully recompile all the BPF progs)

brb replied:
Are you saying that even if a user restarts cilium-agent with different list of `--device`, some programs will run with the old list of devs?

joestringer replied:
I was thinking something more like EKS environments with ENI devices being created / destroyed during a single Cilium agent runtime. Unless we reconfigure these macros, we wouldn't pick up the new devices.

Not sure whether it's possible for such ENIs to be configured and also provide access to the outside world so unless that's a clear use case, it's probably not worth worrying about for now.

joestringer reviewed (commented):
Is commit b61fa714e3da ("datapath: Dynamically determine NATIVE_DEV_IFINDEX") a no-op? Seems like its changes are reverted later in the series. Review was going pretty well up until this point :)

Putting this down for now.

brb committed:
datapath: Add ip-masq-agent implementation

This commit extends the SNAT engine in a way that a packet sent from a
local endpoint to outside is NOT masqueraded if a dst IP addr of the
packet matches the one in the ipmasq map.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
datapath: Use direct routing iface for fwding NodePort requests

In the direct routing mode, when exposing NodePort services on multiple
native devices, we should forward a request to a node running a service
endpoint via a device which is used for direct routing. To make this
happen, we need to know the ifindex of the device (otherwise,
fib_lookup will fail), ipv{4,6} addr of the device for SNAT.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon, datapath: Attach bpf_netdev.o to multiple netdevs

This commit makes it possible to attach bpf_netdev.o to multiple
network devices. Hence, BPF NodePort, BPF MASQ, etc can be exposed
to multiple devices.

Each device can be specified via cilium-agent's --device which accepts
list of strings (e.g. `--device="eth0,eth1"`). The first device in the
list is used to determine a node IP addrs and pod allocation ranges.

With this change, we are no longer to set some macros such
IPV{4,6}_NODEPORT, NATIVE_DEV_IFINDEX, NATIVE_DEV_MAC in node_config.h.

The first two we pass via -D to bpf_netdev.c during the compile time.
While the third one is retrieved via the NATIVE_DEV_MAC_BY_IFINDEX() macro
which is generated by cilium-agent and stored in node_config.h.

Finally, NATIVE_DEV_IFINDEX when bpf_lxc.c does a tail call for the
NodePort rev-DNAT xlation is retrieved from a related CT entry.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
